File: engineeringservicesystem/backend/Constants/Permissions.cs
using System.Collections.Generic;

namespace backend.Constants;

public static class Permissions
{
    // Dashboard
    public const string DashboardView = "Permissions.Dashboard.View";
    
    // Request Views (Granular)
    public const string RequestsView = "Permissions.Requests.View";
    public const string RequestsViewOwn = "Permissions.Requests.ViewOwn";
    public const string RequestsViewBranch = "Permissions.Requests.ViewBranch";
    public const string RequestsViewAll = "Permissions.Requests.ViewAll";
    public const string RequestsViewAssigned = "Permissions.Requests.ViewAssigned";
    
    // Request Actions
    public const string RequestsCreate = "Permissions.Requests.Create";
    public const string RequestsEdit = "Permissions.Requests.Edit";
    public const string RequestsDelete = "Permissions.Requests.Delete";
    public const string RequestsApprove = "Permissions.Requests.Approve";
    public const string RequestsReject = "Permissions.Requests.Reject";
    public const string RequestsAssign = "Permissions.Requests.Assign";
    public const string RequestsManageEngineersWorkLoad = "Permissions.Requests.ManageEngineersWorkLoad";
    public const string RequestsEstimate = "Permissions.Requests.Estimate";

    // User Management
    public const string UserManagementView = "Permissions.UserManagement.View";
    public const string UserManagementManage = "Permissions.UserManagement.Manage";

    // Role Management
    public const string RoleManagementView = "Permissions.RoleManagement.View";
    public const string RoleManagementManage = "Permissions.RoleManagement.Manage";

    // Organization Management
    public const string OrgManagementView = "Permissions.OrgManagement.View";
    public const string OrgManagementManage = "Permissions.OrgManagement.Manage";

    public static List<string> GetAll()
    {
        return new List<string>
        {
            DashboardView,
            RequestsView,
            RequestsViewOwn,
            RequestsViewBranch,
            RequestsViewAll,
            RequestsViewAssigned,
            RequestsCreate,
            RequestsEdit,
            RequestsDelete,
            RequestsApprove,
            RequestsReject,
            RequestsAssign,
            RequestsManageEngineersWorkLoad,
            RequestsEstimate,
            UserManagementView,
            UserManagementManage,
            RoleManagementView,
            RoleManagementManage,
            OrgManagementView,
            OrgManagementManage,
        };
    }
}

File: engineeringservicesystem/frontend/src/constants/permissions.ts
/**
 * Permission Constants - For TypeScript Intellisense Only
 * 
 * ⚠️ IMPORTANT: 
 * - Runtime permission checks use hasPermission('Permissions.XXX')
 * - New permissions added in DB will NOT appear here automatically
 * - This file exists for IDE autocomplete convenience only
 */

export const Permissions = {
    // Dashboard
    DashboardView: "Permissions.Dashboard.View",
    
    // Request Views (Granular)
    RequestsView: "Permissions.Requests.View",
    RequestsViewOwn: "Permissions.Requests.ViewOwn",
    RequestsViewBranch: "Permissions.Requests.ViewBranch",
    RequestsViewAll: "Permissions.Requests.ViewAll",
    RequestsViewAssigned: "Permissions.Requests.ViewAssigned",
    RequestsViewAllEstimated: "Permissions.Requests.ViewAllEstimated",

    // Request Actions
    RequestsCreate: "Permissions.Requests.Create",
    RequestsEdit: "Permissions.Requests.Edit",
    RequestsDelete: "Permissions.Requests.Delete",
    RequestsApprove: "Permissions.Requests.Approve",
    RequestsReject: "Permissions.Requests.Reject",
    RequestsAssign: "Permissions.Requests.Assign",
    RequestsAssignReject: "Permissions.Requests.AssignReject",
    RequestsManageEngineersWorkLoad: "Permissions.Requests.ManageEngineersWorkLoad",
    RequestsEstimate: "Permissions.Requests.Estimate",
    RequestsViewEstimation: "Permissions.Requests.ViewEstimation",
    RequestsViewFilteredEstimation: "Permissions.Requests.ViewFilteredEstimation",
    RequestsSendFilteredEstimation: "Permissions.Requests.SendFilteredEstimation",
    RequestsUploadFinalEstimation: "Permissions.Requests.UploadFinalEstimation",

    // User Management
    UserManagementView: "Permissions.UserManagement.View",
    UserManagementManage: "Permissions.UserManagement.Manage",

    // Role Management
    RoleManagementView: "Permissions.RoleManagement.View",
    RoleManagementManage: "Permissions.RoleManagement.Manage",

    // Organization Management
    OrgManagementView: "Permissions.OrgManagement.View",
    OrgManagementManage: "Permissions.OrgManagement.Manage",

    // Workflow Timeline — one permission controls full timeline visibility
    WorkflowTimelineView: "Permissions.WorkflowTimeline.View",
} as const;

export type PermissionKey = keyof typeof Permissions;